mlflow/pyfunc/backend.py

```diff
@@ -137,33 +230,125 @@ def copy_model_into_container(dockerfile_context_dir):
+                command,
+            ),
+        ],
+        env=command_env,
+    )
 
-        activate_conda_env += [install_mlflow]
-    if os.name != "nt":
-        separator = " && "
-    else:
-        separator = " & "
 
-    command = separator.join(activate_conda_env + [command])
-    _logger.info("=== Running command '%s'", command)
+def should_use_virtualenv():
+    return "MLFLOW_USE_VIRTUALENV" in os.environ
 
-    if os.name != "nt":
-        child = subprocess.Popen(["bash", "-c", command], close_fds=True, env=command_env)
-    else:
-        child = subprocess.Popen(["cmd", "/c", command], close_fds=True, env=command_env)
-    rc = child.wait()
-    if rc != 0:
-        raise Exception(
-            "Command '{0}' returned non zero return code. Return code = {1}".format(command, rc)
+
+def _validate_pyenv_is_available():
+    # TODO
+    pass
+
+
+def _validate_virtualenv_is_available():
+    # TODO
+    pass
+
+
+def _get_env_name_for_virtualenv(requirements_file, python_env_file, env_id=None):
+    requirements_file_contents = open(requirements_file).read()
+    python_env_file_contents = open(python_env_file).read()
+    s = requirements_file_contents + python_env_file_contents + (env_id or "")
+    return "mlflow-" + hashlib.sha1(s.encode("utf-8")).hexdigest()
+
+
+def _execute_in_virtualenv(local_model_path, command, install_mlflow, env_id, command_env):
+    # Read required metadata
+    python_env_file = os.path.join(local_model_path, _PYTHON_ENV_FILE_NAME)
+    python_env = _PythonEnv.from_json(python_env_file)
+
+    before = time.time()
+    _logger.info("Start creating environment")
+
+    # Install python
+    _logger.info("Installing python %s", python_env.python_version)
+    is_windows = _is_windows()
+    # pyenv-win automatically skips the installation
+    pyenv_install_options = () if is_windows else ("--skip-existing",)
+    # pyenv_install_options = ("--force",)
+    _run_command(
+        ["pyenv", "install", *pyenv_install_options, python_env.python_version], command_env
+    )
+
+    # Get the downloaded python location
+    with tempfile.TemporaryDirectory() as tmpdir:
+        # Run `pyenv local` in a temporary directory to avoid creating a `.pyenv-version` file
+        # in the current working directory.
+        python_exec_path = _get_output(
+            [
+                *_get_entrypoint(),
+                _join_commands(
+                    f"pyenv local {python_env.python_version}",
+                    "pyenv which python",
+                ),
+            ],
+            cwd=tmpdir,
+        ).strip()
+    _logger.info("Python executable path: %s", python_exec_path)
+
+    # Generate an environment name
+    home = os.path.expanduser("~")
+    requirements_file = os.path.join(local_model_path, _REQUIREMENTS_FILE_NAME)
+    env_name = _get_env_name_for_virtualenv(requirements_file, python_env_file, env_id)
+
+    # Create an environment
+    env_root = os.path.join(home, ".mlflow", "envs")
```

Comment:
You can also setting special $HOME for `pyenv` to make pyenv isntall python versions into `$HOME/.mlflow/.envs` , like:

`HOME=$HOME/.mlflow/.pythons/ pyenv ...`

Reply:
Are you saying that `HOME=$HOME/.mlflow/.pythons/ pyenv install 3.8.12` install python 3.8.12 in `$HOME/.mlflow/.pythons/`?

Reply:
`HOME=$HOME/.mlflow/.pythons/ pyenv install 3.8.12` It will install into `$HOME/.mlflow/.pythons/.env/versions/3.8.12`

Reply:
In which directory is pyenv installed?

Reply:
I created dockerfile to check this:

```
FROM ubuntu:20.04

RUN apt-get update -y
RUN DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt-get -y install tzdata
RUN apt-get install -y make build-essential libssl-dev zlib1g-dev \
    libbz2-dev wget curl llvm libreadline-dev libsqlite3-dev \
    libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev

RUN wget https://github.com/pyenv/pyenv/archive/refs/tags/v2.2.4.tar.gz -O pyenv.tar.gz \
    && mkdir -p ~/.pyenv \
    && tar -xvf pyenv.tar.gz --strip-components 1 -C ~/.pyenv \
    && rm pyenv.tar.gz
ENV PYENV_ROOT="/root/.pyenv"
ENV PATH="$PYENV_ROOT/bin:$PATH"

RUN mkdir -p /root/.mlflow/.pythons
RUN HOME=/root/.mlflow/.pythons pyenv install 3.7.12
RUN pyenv local 3.7.12
RUN pyenv which python
```

the final line gave:

```
/root/.pyenv/versions/3.7.12/bin/python
```

Reply:
Oh, you should run:
```
HOME=/root/.mlflow/.pythons pyenv local 3.7.12
HOME=/root/.mlflow/.pythons pyenv which python
```

Reply:
Ran the commands you gave:

```
#13 [10/10] RUN HOME=/root/.mlflow/.pythons pyenv which python
#13 sha256:685ae93db2bf583583a5d6673f8edb83fb780d7bd7943edef9c7800e8b99e688
#13 0.252 /root/.pyenv/versions/3.7.12/bin/python
#13 DONE 0.3s
```

Reply:
weird, let me try on ubuntu later.

Reply:
Discussed offline, we can set "PYENV_ROOT" to control the installed place. If it is not set, it will be installed to `$HOME/.pyenv`.

Note: For follow-up NFS optimization, we need to config it to be directory under NFS mounted directory.